File: tests/e2e/test_map_controls_collapse.py
"""
tests/e2e/test_map_controls_collapse.py — Playwright tests for the
collapsible bottom-right map control stack (SNOW-536).

The state machine itself is covered by Vitest against jsdom
(``tests/js/test_map_controls_collapse.js``). What jsdom cannot see is the
part that only exists in a real engine, and that is what this file covers:

1. **The group actually collapses.** jsdom does not compute layout, so the
   CSS height calculation — ``n * --map-control-lg + (n - 1) * 8px``, driven
   off a child count published by the module — is unverifiable there. The
   height being *computed* rather than measured is the whole reason the
   feature works at all (measuring reads ``offsetHeight`` off an
   already-clipped box, which returns 0), so it is worth a real assertion.
2. **The toggle never moves.** The stack is bottom-anchored with the toggle
   as its last child specifically so it stays put while the group grows and
   shrinks above it — a control that slid away under the finger pressing it
   would be the bug this layout exists to avoid.
3. **The preference survives a reload**, i.e. the localStorage round trip
   runs against a real page load and not a re-imported module.
4. **The help tour force-expands a collapsed group.** Several tour steps
   target controls inside the group, and a collapsed control still matches
   ``document.querySelector`` — so ``map_help.js``'s absent-target filter
   does not drop the step, and without the force-expand the highlight ring
   would be positioned against a box clipped to zero height.

Like ``test_map_bottom_bar.py`` these run without a loaded MapLibre style:
the stack is server-rendered and positioned by CSS, and
``map_controls_collapse.js`` only reads ``#map-controls-br``'s DOM. The
ratings endpoint is stubbed so the scrubber reaches its ready state, which
is the signal that the page has finished its boot work.

Transitions are 300ms (``static/css/map.css``), so every assertion that
depends on the animation having finished waits on the settled value rather
than sleeping a fixed amount.
"""

from __future__ import annotations

import json

from playwright.sync_api import Page, Route
from pytest_django.live_server_helper import LiveServer

_STORAGE_KEY = "snowdesk.map.controls.expanded"

# Two dates so sortedDates is non-empty and the scrubber reaches 'ready'
# (mirrors test_map_bottom_bar.py / test_timelapse_popup.py).
_SEASON_RATINGS: str = json.dumps(
    {
        "2026-04-07": {"CH-4115": 2},
        "2026-04-08": {"CH-4115": 3},
    }
)


def _route_season_ratings(route: Route) -> None:
    """Return a stub season-ratings payload so the scrubber enters ready state."""
    route.fulfill(status=200, content_type="application/json", body=_SEASON_RATINGS)


def _navigate_and_wait(page: Page, live_server_url: str) -> None:
    """Navigate to / and wait for the scrubber to signal a finished boot."""
    page.route(f"{live_server_url}/api/ratings/*", _route_season_ratings)
    page.goto(f"{live_server_url}/")
    page.wait_for_load_state("domcontentloaded")
    page.wait_for_selector("#season-scrubber[data-state='ready']", timeout=15_000)


def _group_height(page: Page) -> float:
    """Current rendered height of the collapsible group, in CSS px."""
    return float(
        page.evaluate(
            "() => document.getElementById('map-controls-collapsible')"
            ".getBoundingClientRect().height"
        )
    )


def _toggle_top(page: Page) -> float:
    """Viewport-relative top edge of the collapse toggle, in CSS px."""
    return float(
        page.evaluate(
            "() => document.getElementById('map-controls-toggle')"
            ".getBoundingClientRect().top"
        )
    )


def _wait_for_expanded(page: Page, expanded: bool) -> None:
    """Block until data-expanded settles AND the height transition has finished.

    The attribute flips synchronously on click but the height eases over
    300ms, so asserting on geometry straight after the attribute samples
    mid-flight.

    Waiting for "height > 0" is not enough on the way open — that is true a
    frame into the transition, and an assertion on the final height then sees
    a partial value (this test first failed at 24.9px against an expected
    184px). So require the height to be *stable* across consecutive polls —
    Playwright polls wait_for_function on animation frames — and only then
    check it against the direction we asked for.
    """
    page.wait_for_selector(
        f"#map-controls-br[data-expanded='{str(expanded).lower()}']", timeout=5_000
    )
    # Reset the sampler so a value left over from an earlier wait can't be
    # mistaken for a settled reading on the first poll.
    page.evaluate("() => { delete window.__lastGroupHeight; }")
    page.wait_for_function(
        """(wantOpen) => {
            const h = document.getElementById('map-controls-collapsible')
                .getBoundingClientRect().height;
            const prev = window.__lastGroupHeight;
            window.__lastGroupHeight = h;
            // Still moving between frames — keep polling.
            if (prev === undefined || Math.abs(h - prev) > 0.5) return false;
            return wantOpen ? h > 1 : h < 1;
        }""",
        arg=expanded,
        timeout=5_000,
    )


def test_collapsing_hides_the_group_without_moving_the_toggle(
    live_server: LiveServer,
    page: Page,
    _load_test_data: None,
) -> None:
    """Collapsing clears the group's height and leaves the toggle where it was.

    The stack is bottom-anchored with the toggle last, so it grows and
    shrinks upward: the toggle must not move under the finger pressing it.
    """
    page_errors: list[str] = []
    page.on("pageerror", lambda err: page_errors.append(str(err)))
    _navigate_and_wait(page, live_server.url)

    # Default is expanded, and the computed height is real — this is the
    # assertion jsdom cannot make.
    _wait_for_expanded(page, True)
    expanded_height = _group_height(page)
    assert expanded_height > 0, (
        "the collapsible group should have a computed open height; "
        f"got {expanded_height}px"
    )

    toggle_top_before = _toggle_top(page)

    page.locator("#map-controls-toggle").click()
    _wait_for_expanded(page, False)

    assert _group_height(page) < 1, (
        f"the group should collapse to zero height; got {_group_height(page)}px"
    )
    # The whole point of anchoring the toggle last: it stays put.
    assert abs(_toggle_top(page) - toggle_top_before) <= 1, (
        "the toggle must not move when the group collapses; "
        f"moved from {toggle_top_before}px to {_toggle_top(page)}px"
    )

    # Re-expanding restores the same computed height.
    page.locator("#map-controls-toggle").click()
    _wait_for_expanded(page, True)
    assert abs(_group_height(page) - expanded_height) <= 1, (
        "re-expanding should restore the original height; "
        f"{_group_height(page)}px vs {expanded_height}px"
    )

    assert page_errors == [], f"JS errors: {page_errors}"


def test_collapsed_preference_survives_a_reload(
    live_server: LiveServer,
    page: Page,
    _load_test_data: None,
) -> None:
    """Collapsing persists to localStorage and is restored on the next load."""
    _navigate_and_wait(page, live_server.url)
    _wait_for_expanded(page, True)

    page.locator("#map-controls-toggle").click()
    _wait_for_expanded(page, False)

    stored = page.evaluate(f"() => localStorage.getItem({_STORAGE_KEY!r})")
    assert stored == "false", (
        f"expected the collapsed preference to persist; got {stored!r}"
    )

    _navigate_and_wait(page, live_server.url)
    _wait_for_expanded(page, False)
    assert _group_height(page) < 1, (
        "a stored collapsed preference should be applied on the next load"
    )
    assert (
        page.locator("#map-controls-toggle").get_attribute("aria-label")
        == "Show map controls"
    ), "the toggle label should describe what the next tap does"


def test_help_tour_force_expands_a_collapsed_group(
    live_server: LiveServer,
    page: Page,
    _load_test_data: None,
) -> None:
    """Opening the tour expands a collapsed group, and closing it restores.

    Tour steps target controls inside the group. A collapsed control still
    matches ``document.querySelector``, so map_help.js's absent-target filter
    keeps the step and the ring would otherwise land on a clipped box. The
    user's stored preference must not be overwritten by the tour.
    """
    _navigate_and_wait(page, live_server.url)
    _wait_for_expanded(page, True)

    page.locator("#map-controls-toggle").click()
    _wait_for_expanded(page, False)

    # The "?" roundel lives inside the collapsed group, so drive the tour the
    # way an auto-start would rather than clicking a clipped control.
    page.evaluate(
        "() => document.dispatchEvent(new CustomEvent('snowdesk:map-help-open'))"
    )
    _wait_for_expanded(page, True)
    assert page.evaluate(f"() => localStorage.getItem({_STORAGE_KEY!r})") == "false", (
        "the tour must not overwrite the user's stored preference"
    )

    page.evaluate(
        "() => document.dispatchEvent(new CustomEvent('snowdesk:map-help-close'))"
    )
    _wait_for_expanded(page, False)
    assert _group_height(page) < 1, (
        "closing the tour should hand the group back to the stored preference"
    )

File: tests/js/test_map_controls_collapse.js
/*
 * tests/js/test_map_controls_collapse.js — Vitest unit tests for
 * static/js/map_controls_collapse.js.
 *
 * The module reads #map-controls-br / #map-controls-collapsible /
 * #map-controls-toggle and returns early if any is absent, so the fixture
 * (mirroring apps/public/templates/public/partials/_map_embed.html) must exist
 * BEFORE the module is imported. Each test re-imports via vi.resetModules()
 * so the IIFE runs fresh against its own fixture and localStorage state.
 *
 * Only state is asserted here — data-expanded, ARIA, the persisted
 * preference and the --map-controls-count custom property. The open/close
 * animation is pure CSS (static/css/map.css) and jsdom does not compute it;
 * the height it produces is covered by browser-side checks instead.
 */

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import '../../static/js/i18n_strings.js';

const STORAGE_KEY = 'snowdesk.map.controls.expanded';

/**
 * Build the control stack: layers + locate outside the group, four
 * collapsible controls inside it, toggle anchored last.
 *
 * @param {number} collapsibleCount - children rendered inside the group.
 */
function buildFixture(collapsibleCount = 4) {
  const items = Array.from(
    { length: collapsibleCount },
    (_, i) => `<div class="map-utility-pill" id="collapsible-${i}"></div>`,
  ).join('');
  document.body.innerHTML = `
    <div class="map-controls-br" id="map-controls-br" data-expanded="true">
      <div class="map-utility-pill map-utility-pill--basemap" id="basemap-pill"></div>
      <div class="map-utility-pill map-utility-pill--locate"></div>
      <div id="map-controls-collapsible">
        <div class="map-controls-collapsible-inner">${items}</div>
      </div>
      <button type="button" id="map-controls-toggle"
              aria-controls="map-controls-collapsible" aria-expanded="true"></button>
    </div>
  `;
}

async function loadModule() {
  vi.resetModules();
  await import('../../static/js/map_controls_collapse.js');
}

const stack = () => document.getElementById('map-controls-br');
const toggle = () => document.getElementById('map-controls-toggle');

beforeEach(() => {
  localStorage.clear();
});

afterEach(() => {
  document.body.innerHTML = '';
  localStorage.clear();
});

describe('map controls collapse', () => {
  it('defaults to expanded when nothing is stored', async () => {
    buildFixture();
    await loadModule();
    expect(stack().dataset.expanded).toBe('true');
    expect(toggle().getAttribute('aria-expanded')).toBe('true');
    expect(toggle().getAttribute('aria-label')).toBe('Hide map controls');
  });

  it('publishes the live child count for the CSS height calculation', async () => {
    buildFixture(3);
    await loadModule();
    const group = document.getElementById('map-controls-collapsible');
    // Read from the DOM, not hardcoded — a flag-gated control must not leave
    // the group clipped short or standing open too tall.
    expect(group.style.getPropertyValue('--map-controls-count')).toBe('3');
  });

  it('collapses on click and persists the choice', async () => {
    buildFixture();
    await loadModule();
    toggle().click();
    expect(stack().dataset.expanded).toBe('false');
    expect(toggle().getAttribute('aria-expanded')).toBe('false');
    expect(toggle().getAttribute('aria-label')).toBe('Show map controls');
    expect(localStorage.getItem(STORAGE_KEY)).toBe('false');
  });

  it('restores a stored collapsed preference on load', async () => {
    localStorage.setItem(STORAGE_KEY, 'false');
    buildFixture();
    await loadModule();
    expect(stack().dataset.expanded).toBe('false');
    expect(toggle().getAttribute('aria-label')).toBe('Show map controls');
  });

  it('forces the group open for the help tour without touching the preference', async () => {
    localStorage.setItem(STORAGE_KEY, 'false');
    buildFixture();
    await loadModule();
    expect(stack().dataset.expanded).toBe('false');

    // Steps target controls inside the group; a collapsed one still matches
    // querySelector, so the ring would land on a zero-height box.
    document.dispatchEvent(new CustomEvent('snowdesk:map-help-open'));
    expect(stack().dataset.expanded).toBe('true');
    expect(localStorage.getItem(STORAGE_KEY)).toBe('false');

    document.dispatchEvent(new CustomEvent('snowdesk:map-help-close'));
    expect(stack().dataset.expanded).toBe('false');
  });

  it('leaves an already-expanded group alone across the tour', async () => {
    buildFixture();
    await loadModule();
    document.dispatchEvent(new CustomEvent('snowdesk:map-help-open'));
    expect(stack().dataset.expanded).toBe('true');
    // Not forced open, so closing the tour must not collapse it.
    document.dispatchEvent(new CustomEvent('snowdesk:map-help-close'));
    expect(stack().dataset.expanded).toBe('true');
  });

  it('adopts collapsed when tapped while the tour holds it open', async () => {
    localStorage.setItem(STORAGE_KEY, 'false');
    buildFixture();
    await loadModule();
    document.dispatchEvent(new CustomEvent('snowdesk:map-help-open'));
    expect(stack().dataset.expanded).toBe('true');

    // Without this the tap would appear to do nothing (already open) and the
    // group would silently revert when the tour closed.
    toggle().click();
    expect(stack().dataset.expanded).toBe('false');
    expect(localStorage.getItem(STORAGE_KEY)).toBe('false');

    document.dispatchEvent(new CustomEvent('snowdesk:map-help-close'));
    expect(stack().dataset.expanded).toBe('false');
  });

  it('is inert when the stack is absent', async () => {
    document.body.innerHTML = '<div id="unrelated"></div>';
    await expect(loadModule()).resolves.not.toThrow();
  });
});

describe('translated labels (SNOW-620)', () => {
  /**
   * Append a strings template to the fixture, as _map_embed.html does.
   *
   * @param {Object<string, string>} entries
   */
  function mountStrings(entries) {
    const tpl = document.createElement('template');
    tpl.id = 'map-controls-strings-template';
    tpl.innerHTML = Object.entries(entries)
      .map(([key, value]) => `<span data-string="${key}">${value}</span>`)
      .join('');
    document.body.appendChild(tpl);
  }

  it('labels the toggle from the strings template when it is present', async () => {
    buildFixture();
    mountStrings({ collapse: 'Bedienelemente ausblenden', expand: 'Bedienelemente' });
    await loadModule();

    // Rendered expanded, so the label describes what the next tap does.
    expect(toggle().getAttribute('aria-label')).toBe('Bedienelemente ausblenden');

    toggle().click();

    expect(toggle().getAttribute('aria-label')).toBe('Bedienelemente');
  });

  it('falls back to English when the partial did not render the template', async () => {
    // The real case this guards: a page that includes the module but not
    // the surface partial. The control must still be labelled.
    buildFixture();
    await loadModule();

    expect(toggle().getAttribute('aria-label')).toBe('Hide map controls');

    toggle().click();

    expect(toggle().getAttribute('aria-label')).toBe('Show map controls');
  });
});
